File: app/src/test/java/com/codeprogression/mvpb/MainPresenterTest.java
package com.codeprogression.mvpb;

import org.junit.Before;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class MainPresenterTest {

    private MainPresenter presenter;
    private MainView.ViewModel viewModel;

    @Before
    public void setup(){
        presenter = new MainPresenter();
        viewModel = new MainView.ViewModel();
    }

    @Test
    public void smokeTest(){
        presenter.attach(viewModel);
        assertThat(viewModel.number.get()).isEqualTo(10);
    }

    @Test
    public void test_presenterUpdatesViewModel(){
        presenter.attach(viewModel);
        presenter.add();
        assertThat(viewModel.number.get()).isEqualTo(11);
    }

    @Test
    public void test_presenterResetsNumber(){
        viewModel.number.set(1);
        presenter.attach(viewModel);
        assertThat(viewModel.number.get()).isEqualTo(10);
    }

    @Test
    public void test_presenterDoesNotResetNumber(){
        viewModel.number.set(15);
        presenter.attach(viewModel);
        assertThat(viewModel.number.get()).isEqualTo(15);
    }
}

File: app/src/test/java/com/codeprogression/mvpb/MainPresenterTest.kt
package com.codeprogression.mvpb

import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test

class MainPresenterTest {

    lateinit private var presenter: MainPresenter
    lateinit private var viewModel: MainView.ViewModel

    @Before
    fun setup() {
        presenter = MainPresenter()
        viewModel = MainView.ViewModel()
    }

    @Test
    fun smokeTest() {
        presenter.attach(viewModel)
        assertThat(viewModel.number.get()).isEqualTo(10)
    }

    @Test
    fun test_presenterUpdatesViewModel() {
        presenter.attach(viewModel)
        presenter.add()
        assertThat(viewModel.number.get()).isEqualTo(11)
    }

    @Test
    fun test_presenterResetsNumber() {
        viewModel.number.set(1)
        presenter.attach(viewModel)
        assertThat(viewModel.number.get()).isEqualTo(10)
    }

    @Test
    fun test_presenterDoesNotResetNumber() {
        viewModel.number.set(15)
        presenter.attach(viewModel)
        assertThat(viewModel.number.get()).isEqualTo(15)
    }
}